src/ucs/sys/rcache.c

```diff
@@ -442,12 +447,19 @@ ucs_rcache_create_region(ucs_rcache_t *rcache, void *address, size_t length,
         UCS_PROFILE_NAMED_CALL("mem_reg", rcache->params.ops->mem_reg,
                                rcache->params.context, rcache, arg, region);
     if (status != UCS_OK) {
-        /* In case region is not registered, we don't return it to the user,
-         * so no need to increment reference count.
-         */
-        ucs_rcache_region_debug(rcache, region, "created with status %s",
-                                ucs_status_string(status));
-        goto out_unlock;
+        if (merged) {
+            /* failure may be due to merge, retry with original address */
+            ucs_debug("failed to register merged region " UCS_PGT_REGION_FMT ": %s, retrying",
+                      UCS_PGT_REGION_ARG(&region->super), ucs_status_string(status));
+            status = ucs_pgtable_remove(&rcache->pgtable, &region->super);
+            ucs_assert_always(status == UCS_OK);
+            ucs_mem_region_destroy_internal(rcache, region);
+            goto retry;
```

Comment:
no, because the next iteration 'merge' flag would be 0 - the previous overlapping regions are already destroyed, and no new regions could be created because we are under lock